File: packages/ts/types/src/reaction.ts
import { z } from 'zod';
import { BaseEntitySchema } from './common.js';

export const ReactionKindSchema = z.enum(['unicode', 'custom_emoji']);
export type ReactionKind = z.infer<typeof ReactionKindSchema>;

export const MessageReactionSchema = BaseEntitySchema.extend({
	message_id: z.string(),
	user_id: z.string(),
	kind: ReactionKindSchema,
	value: z.string().describe('Emoji unicode char or custom emoji shortcode'),
	image_url: z.string().url().optional().describe('URL for custom emoji image')
});

export type MessageReaction = z.infer<typeof MessageReactionSchema>;

File: packages/rust/syren-types/src/reaction.rs
//! Reaction (emoji) shapes.

use serde::{Deserialize, Serialize};
use zod_gen_derive::ZodSchema;

#[cfg(target_arch = "wasm32")]
use tsify_next::Tsify;

#[derive(Clone, Debug, Serialize, Deserialize, ZodSchema)]
#[cfg_attr(target_arch = "wasm32", derive(Tsify))]
#[cfg_attr(target_arch = "wasm32", tsify(into_wasm_abi, from_wasm_abi))]
pub enum ReactionKind {
	#[serde(rename = "unicode")]
	Unicode,
	#[serde(rename = "custom_emoji")]
	CustomEmoji,
}

#[derive(Clone, Debug, Serialize, Deserialize, ZodSchema)]
#[cfg_attr(target_arch = "wasm32", derive(Tsify))]
#[cfg_attr(target_arch = "wasm32", tsify(into_wasm_abi, from_wasm_abi))]
pub struct MessageReaction {
	pub id: String,
	pub message_id: String,
	pub user_id: String,
	pub kind: ReactionKind,
	pub value: String,
	#[serde(skip_serializing_if = "Option::is_none", default)]
	pub image_url: Option<String>,
	pub created_at: String,
	pub updated_at: String,
}
